controller/node_controller.go

```diff
@@ -1018,3 +1001,215 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
 		delete(bi.Spec.Disks, diskUUID)
 	}
 }
+
+func (nc *NodeController) checkMonitor(node *longhorn.Node) (monitor.Monitor, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	if nc.monitor != nil {
+		return nc.monitor, nil
+	}
+
+	monitor, err := monitor.NewNodeMonitor(nc.logger, nc.ds, node, nc.enqueueNodeForMonitor)
+	if err != nil {
+		return nil, err
+	}
+
+	nc.monitor = monitor
+
+	return monitor, nil
+}
+
+func (nc *NodeController) enqueueNodeForMonitor(key string) {
+	nc.queue.Add(key)
+}
+
+func (nc *NodeController) syncOrphans(node *longhorn.Node, collectedDataInfo map[string]*monitor.CollectedDiskInfo) error {
+	for diskName, diskInfo := range collectedDataInfo {
+		if err := nc.createOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to create orphans for disk %v", diskName)
+		}
+		if err := nc.deleteOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to delete orphans for disk %v", diskName)
+		}
+	}
+
+	return nil
+}
+
+func (nc *NodeController) deleteOrphans(node *longhorn.Node, diskName string, diskInfo *monitor.CollectedDiskInfo) error {
+	autoDeletionEnabled, err := nc.ds.GetSettingAsBool(types.SettingNameOrphanAutoDeletion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameOrphanAutoDeletion)
+	}
+
+	orphans, err := nc.ds.ListOrphansRO()
+	if err != nil {
+		return errors.Wrap(err, "unable to list orphans")
+	}
+
+	for _, orphan := range orphans {
+		if orphan.Status.OwnerID != nc.controllerID {
+			continue
+		}
```

Comment:
This implicitly means that Longhorn won't automatically delete the Orphan CRs on the down/removed nodes, right?

Reply:
The `OwnerID` of an orphan CR, whose node is down or removed, will be updated to another NodeID.

The orphan controller checks if the orphan CR' node is missing or not. If missing, `DataCleanable` condition is set to `False`, node controller will clean it up.

`if orphan.Status.OwnerID != nc.controllerID {` is to ensure and prevent from the mistaken deletion of data.

Reply:
I'm thinking this logic is redundant.

Reply:
The condition `DataCleanable` would tell if the data can be removed or not. It won't help the current node controller skip the handling of the orphans on other nodes...

Reply:
> If yes, why not directly use ListOrphansByNodeRO?

The label is still equal to the `orphan.spec.NodeID`, so I don't use `ListOrphansByNodeRO` here.

Reply:
You mean this check is already included in `isOnDiskDataMissing`? If there are limited Orphan CRs in the cluster, I am fine with either. Otherwise, I prefer to filter them at the beginning (via the ListOption).

Reply:
yeah, I also prefer to filter them first.

Reply:
Addressed.